source/2_platform/rex_directx/src/directx_renderer.cpp

```diff
@@ -73,76 +238,1816 @@ namespace rex
+          sd.BufferDesc.RefreshRate.Numerator   = userData.refresh_rate;
+          sd.BufferDesc.RefreshRate.Denominator = 1;
+          sd.BufferDesc.Format                  = g_ctx.back_buffer_format;
+          sd.BufferDesc.ScanlineOrdering        = DXGI_MODE_SCANLINE_ORDER_UNSPECIFIED;
+          sd.BufferDesc.Scaling                 = DXGI_MODE_SCALING_UNSPECIFIED;
+          sd.SampleDesc.Count                   = g_ctx.msaa_state ? 4 : 1;
+          sd.SampleDesc.Quality                 = g_ctx.msaa_state ? g_ctx.msaa_quality - 1 : 0;
+          sd.BufferUsage                        = DXGI_USAGE_RENDER_TARGET_OUTPUT;
+          sd.BufferCount                        = s_swapchain_buffer_count;
+          sd.OutputWindow                       = (HWND)userData.primary_display_handle;
+          sd.Windowed                           = userData.windowed;
+          sd.SwapEffect                         = DXGI_SWAP_EFFECT_FLIP_DISCARD;
+          sd.Flags                              = DXGI_SWAP_CHAIN_FLAG_ALLOW_MODE_SWITCH;
+
+          // Note: swap chain uses queue to perform flush.
+          rex::wrl::com_ptr<IDXGIFactory> dxgi_factory = factory->as<IDXGIFactory>();
+          if(FAILED(dxgi_factory->CreateSwapChain(g_ctx.command_queue.Get(), &sd, g_ctx.swapchain.GetAddressOf())))
+          {
+            REX_ERROR(LogDirectX, "Failed to create swap chain");
+            return false;
+          }
+
+          directx::set_debug_name_for(g_ctx.swapchain.Get(), "SwapChain");
+
+          return true;
+        }
+
+        //-------------------------------------------------------------------------
+        wrl::com_ptr<ID3D12Resource> create_default_buffer(ID3D12Device* device, ID3D12GraphicsCommandList* cmdList, const void* initData, UINT64 byteSize, wrl::com_ptr<ID3D12Resource>& uploadBuffer)
+        {
+          wrl::com_ptr<ID3D12Resource> default_buffer;
+
+          // Create the actual default buffer resource.
+          CD3DX12_HEAP_PROPERTIES heap_properties_default(D3D12_HEAP_TYPE_DEFAULT);
+          auto buffer_default = CD3DX12_RESOURCE_DESC::Buffer(byteSize);
+
+          if(FAILED(device->CreateCommittedResource(&heap_properties_default, D3D12_HEAP_FLAG_NONE, &buffer_default, D3D12_RESOURCE_STATE_COMMON, nullptr, IID_PPV_ARGS(default_buffer.GetAddressOf()))))
+          {
+            REX_ERROR(LogDirectX, "Failed to create commited resource for a default buffer.");
+            return nullptr;
+          }
+
+          // In order to copy CPU memory data into our default buffer, we need to create
+          // an intermediate upload heap.
+          CD3DX12_HEAP_PROPERTIES heap_properties_upload(D3D12_HEAP_TYPE_UPLOAD);
+          auto buffer_upload = CD3DX12_RESOURCE_DESC::Buffer(byteSize);
+
+          if(FAILED(device->CreateCommittedResource(&heap_properties_upload, D3D12_HEAP_FLAG_NONE, &buffer_upload, D3D12_RESOURCE_STATE_GENERIC_READ, nullptr, IID_PPV_ARGS(uploadBuffer.GetAddressOf()))))
```

Comment:
you're creating an upload buffer for every buffer?
I think it's best to create 1 (or a few) upload buffers, and then reuse those, every time you want to upload data to the GPU

Reply:
This is somewhat of an overhaul on the creation of upload buffers, I agree we should not create an upload buffer for every buffer ( although I read through documentation and DX12 book and seems like this is a standard design ... at least I can try if it works). It would be consistent with how other objects work (eg: RTV, DSV, CBV ). But I will address this in another PR.

Reply:
linked to https://github.com/New-Creta/Rex/issues/251